### PR TITLE
Open Telemetry

### PR DESCRIPTION
**Pull request by Xaelias**
This PR implements otel for http and thrift in bp.py.
Because of unsolvable conflicting dependencies, this also removes python 3.7.

### CODE REVIEW

**Comment by trevorriles** [comment]
@chriskuehl you mentioned that there is some changes that have made it to develop but not been released yet. Would it be better for us to target a release branch?

File: .github/workflows/python-package.yaml
Comment (chriskuehl) [inline .github/workflows/python-package.yaml]: Looks like we duplicated some of the Python 3.7 dropping work; I made similar changes in #823 (isolated in the first commit).

Maybe it makes sense for us to pull that into a separate PR? TODO for me.

File: Dockerfile
Comment (chriskuehl) [inline Dockerfile]: Do you recall what package needed this? I wonder if there's a possibility of choosing a version that has manylinux wheels uploaded to avoid needing to compile it ourselves at install time.

File: baseplate/clients/requests.py
Comment (chriskuehl) [inline baseplate/clients/requests.py:29]: It looks like the way these work is essentially to patch these libraries to inject tracing logic. I don't think that's inherently a problem but it's definitely a little sneaky, especially to do as an import-time side effect.

In general I think it's best if importing code doesn't have any side effects. This matches people's intuition about what importing does, reduces weird surprises (e.g. adding a seemingly-unrelated import suddenly changes behavior), is faster, and plays better with linters and any other tools that need to import code. It also makes it possible to access any configuration the app provides so that you can adjust the instrumentation settings if needed (e.g. if we wanted to add a way to toggle off instrumentation, or make it possible to exempt certain routes).

Do you think there's a way we can move this out of the import path and instead have it happen at startup time?

File: baseplate/clients/thrift.py
Comment (chriskuehl) [inline baseplate/clients/thrift.py]:
```suggestion
        except socket.gaierror:
```

File: baseplate/clients/thrift.py
Comment (chriskuehl) [inline baseplate/clients/thrift.py]:
```suggestion
                        except BaseException:
```

Explicitly writing `except BaseException:` is the way to avoid the linter error when that's your goal.

File: baseplate/frameworks/pyramid/__init__.py
Comment (chriskuehl) [inline baseplate/frameworks/pyramid/__init__.py:41]: Same question as before about instrumenting as an import-time side effect.

File: baseplate/frameworks/pyramid/__init__.py
Comment (chriskuehl) [inline baseplate/frameworks/pyramid/__init__.py]: This is a behavior change (bare except catches all exceptions, `except Exception` catches only exceptions that inherit from Exception. So, this no longer catches exceptions like `KeyboardInterrupt`. Is that OK? If you want to keep the old behavior, you can use `except BaseException as e` instead, which is equivalent to the bare except.

You should be able to drop the `noqa` here now that you're not using a bare except.

File: baseplate/frameworks/pyramid/__init__.py
Comment (chriskuehl) [inline baseplate/frameworks/pyramid/__init__.py]: Ditto

File: baseplate/frameworks/thrift/__init__.py
Comment (chriskuehl) [inline baseplate/frameworks/thrift/__init__.py]: Probably can drop noqa, same comments as before.

File: baseplate/lib/propagator_redditb3.py
Comment (chriskuehl) [inline baseplate/lib/propagator_redditb3.py]:
```suggestion
    _SAMPLE_PROPAGATE_VALUES = frozenset(("1", "True", "true", "d"))
```

File: baseplate/lib/propagator_redditb3.py
Comment (chriskuehl) [inline baseplate/lib/propagator_redditb3.py]: Since there's no anchor at the end, I think it will always just select the first 16 chars even for a longer ID?

```python
>>> re.match(r"[\da-fA-F]{16}|[\da-fA-F]{32}", "aaaabbbbccccddddaaaabbbbccccdddd").group(0)
'aaaabbbbccccdddd'
```

It also allows other character lengths as long as they're at least 16. Not sure exactly what the intention of the regex is here but given that the `|[\da-fA-F]{32}` part isn't doing anything, this seems unexpected?

**Review by chriskuehl — COMMENTED** [review]
First pass review -- it would be a good idea to have someone more familiar with the tracing bits also take a look.

File: .github/workflows/python-package.yaml
Comment (chriskuehl) [reply .github/workflows/python-package.yaml]: Extracted to https://github.com/reddit/baseplate.py/pull/824

File: Dockerfile
Comment (Xaelias) [reply Dockerfile]: I am not 100% sure anymore but I think it had smth to do with the apple silicon MBPs.

Comment (chriskuehl) [reply Dockerfile]: Ah, interesting. I actually changed it to use x86_64 emulation in my PR (https://github.com/reddit/baseplate.py/pull/823/commits/97ff89d5bca69405e016d799c742fcb5ec697c6a) since there was no Thrift compiler image for ARM and it was failing on my MacBook trying to build a native arch version of the image. This is probably fine to keep as-is for now, though I'm surprised you didn't run into the same Thrift issue I did (maybe I'm doing something wrong?).

Comment (Xaelias) [reply Dockerfile]: ```DOCKER_DEFAULT_PLATFORM=linux/amd64 docker-compose build```
worked locally for me

Comment (Xaelias) [reply Dockerfile]: so two different paths to achieve somewhat similar results :-D

File: baseplate/clients/requests.py
Comment (Xaelias) [reply baseplate/clients/requests.py:29]: Honestly I don't know. I think it would make testing much more complex since you'd need to basically load the entire app to test anything.
It's also how otel says you should be using these auto-instrumenter.

Comment (KTAtkinson) [reply baseplate/clients/requests.py:29]: I agree with Chris that this smells a little funny. I don't see anything in their API docs that suggests that this needs to be instrumented the top level like this. Can you share the link to the instructions/example you are following?

**Review by KTAtkinson — CHANGES_REQUESTED** [review]

File: baseplate/clients/requests.py
Comment (Xaelias) [reply baseplate/clients/requests.py:29]: Can you check the commit I just pushed? Is that closer to what y'all had in mind? (we change names etc. I just wanted to get smth out there for you to be able to look at)

**Comment by trevorriles** [comment]
I've got develop merged in and all the tests passing. I'll begin addressing the other PR comments next.

One thing to note is that opentelemetry-python doesn't officially support 3.12 yet, so we may have to wait for that or restrict baseplate to 3.11 for the time being.

File: baseplate/lib/propagator_redditb3.py
Comment (trevorriles) [reply baseplate/lib/propagator_redditb3.py]: I think your correct here, I lifted this from upstream here and modified it for Reddit's format.
https://github.com/open-telemetry/opentelemetry-python/blob/d6321d627cd0fc9345a689a9a37cf4cb9ea44ba0/propagator/opentelemetry-propagator-b3/src/opentelemetry/propagators/b3/__init__.py#L46-L47

Let me dig in a bit more to make sure we get the best behavior.

File: baseplate/frameworks/pyramid/__init__.py
Comment (trevorriles) [reply baseplate/frameworks/pyramid/__init__.py]: In this case I think we are ok using `Exception`. The opentelemetry lib want's an `Exception` for it's `record_exception` function. As in all these cases we just end the span and re-raise, I think it's fine if we miss ending the span for exceptions that inherit from `BaseException` but not `Exception`.

File: baseplate/lib/propagator_redditb3.py
Comment (trevorriles) [reply baseplate/lib/propagator_redditb3.py]: Ahh it uses the [re.fullmatch](https://docs.python.org/3/library/re.html#re.fullmatch) function, which if I read it correctly means it will check the second option.

Comment (trevorriles) [reply baseplate/lib/propagator_redditb3.py]: Either way this is incorrect as the baseplate spec uses base10 encoded integers... I'll get a fix together.

File: baseplate/clients/requests.py
Comment (trevorriles) [reply baseplate/clients/requests.py:29]: I think @Xaelias 's change works for me what are your thoughts @KTAtkinson & @chriskuehl ?

File: baseplate/lib/propagator_redditb3.py
Comment (trevorriles) [reply baseplate/lib/propagator_redditb3.py]: Got the fix in to look for base10 integers wrapped in a string.

**Comment by trevorriles** [comment]
@chriskuehl I'm curious what your thoughts are on the lack of 3.12 support at the moment?

https://github.com/open-telemetry/opentelemetry-python/issues/3617

**Comment by chriskuehl** [comment]
> @chriskuehl I'm curious what your thoughts are on the lack of 3.12 support at the moment?

I don't think we can merge this until 3.12 support is available, since we already have services which have migrated to 3.12.

**Review by KTAtkinson — APPROVED** [review]

File: Dockerfile
Comment (chriskuehl) [inline Dockerfile]: Can we move this new apt line above the COPY so that we don't need to reinstall apt packages every time a requirement changes? Maybe it can go directly under the `FROM python:3.12` line?

File: baseplate/clients/thrift.py
Comment (chriskuehl) [inline baseplate/clients/thrift.py:276]: When does localhost traffic like this happen? For the IPv4 address I guess in theory it could be all of 127.0.0.0/8 depending on how this is used? Mostly wondering because it's not uncommon to see e.g. `127.0.1.1` used in default Linux setups in `/etc/hosts`.

File: baseplate/lib/propagator_redditb3_http.py
Comment (chriskuehl) [inline baseplate/lib/propagator_redditb3_http.py]:
```suggestion
    _id_regex = re_compile(r"\d+")
```

File: baseplate/lib/propagator_redditb3_thrift.py
Comment (chriskuehl) [inline baseplate/lib/propagator_redditb3_thrift.py]:
```suggestion
    _id_regex = re_compile(r"\d+")
```

File: baseplate/server/__init__.py
Comment (chriskuehl) [inline baseplate/server/__init__.py]:
```suggestion
            if config.tracing.get("insecure") == "false":
                insecure = False
            else:
                # Default to true. At the time of implementation our current tracing pipeline is not setup for TLS.
                insecure = True
```

File: tests/integration/otel_thrift_tests.py
Comment (chriskuehl) [inline tests/integration/otel_thrift_tests.py]: How does this exception happen? Is it possible to catch a specific error?

**Review by chriskuehl — APPROVED** [review]
This looks reasonable but given the size it's difficult to catch any bugs during review. I think we should discuss how we want to test this before the final release.

Let's also get a review from someone more familiar with our tracing setup before merging, since I'm not really qualified to review that beyond basic logic checking.

File: Dockerfile
Comment (trevorriles) [reply Dockerfile]: Good call, I'll move this up.

File: baseplate/clients/thrift.py
Comment (trevorriles) [reply baseplate/clients/thrift.py:276]: Honestly I think the effort here is to just provide a default `NET_PEER_NAME` for some common scenarios. It's kind of a convenience if you are filtering or grouping by hostnames I guess so that you get a value instead of empty.

File: tests/integration/otel_thrift_tests.py
Comment (trevorriles) [reply tests/integration/otel_thrift_tests.py]: I'll look into the exceptions thrown here and clean this up.

**Comment by trevorriles** [comment]
> This looks reasonable but given the size it's difficult to catch any bugs during review. I think we should discuss how we want to test this before the final release.

Fully on board here, here are two testing steps I want to take before we cut an official release:
- [ ] Additional testing in snoodev to verify context propagation general functionality
- [ ] Cut some form of alpha release and work with teams that are still < 3.12. This will make it easier for them to test in snoodev and possibly even some canary testing.

With the 3.12 limitation I'm interested in your thoughts around maintaining an otel branch/release that will be limited to 3.11 and below until we can get the upstream issues resolved?

> Let's also get a review from someone more familiar with our tracing setup before merging, since I'm not really qualified to review that beyond basic logic checking.

I'm sure we can get @Xaelias to give this another pass, since I've taken this over. Not sure if you can approve your own PR though. Ultimately we are the two people who have context to the tracing work here. Pete Loaf has done most of the golang work, but I'm not sure if it makes sense to have him review here as well or not?

File: baseplate/server/__init__.py
Comment (trevorriles) [inline baseplate/server/__init__.py]: I need to look at moving these back.

This created an infinite loop somehow related to the import timing of the `ssl` lib and our gevent `patch_all()` call in `baseplate-serve`. 

Honestly I don't 100% know why yet, still digging.

Comment (trevorriles) [reply baseplate/server/__init__.py]: Or an alternative that is not this 😓

File: tests/integration/otel_thrift_tests.py
Comment (trevorriles) [reply tests/integration/otel_thrift_tests.py]: This method wasn't called so I just removed it.

I also removed it from the `thrift_tests.py` file as well as that's where we lifted most of these tests.

**Review by trevorriles — APPROVED** [review]

File: baseplate/server/__init__.py
Comment (trevorriles) [inline baseplate/server/__init__.py]: @chriskuehl I'm trying to use `DefaultFromEnv` as suggested but I think I'm misunderstanding how this is supposed to work. Does this also require implementing these options in the ini based config to get it to work?

File: baseplate/frameworks/thrift/__init__.py
Comment (redloaf) [inline baseplate/frameworks/thrift/__init__.py]: 🔕 I believe this formats the log message even if debug logging is disabled.
```suggestion
            logger.debug("Extracted trace headers. [ctx=%s, header_dict=%s]", ctx, header_dict)
```

**Review by redloaf — APPROVED** [review]

**Pull request merged**
